Attestation/ek-secret.md

```diff
@@ -0,0 +1,173 @@
+# `Passing a secret to a TPM using only the EK`
+
+This is example code to pass a secret to a system by just knowing the endorsenment key.
+We will be using the current (commit 07a92e9fa75548ea102ce90b3b6182093b3f7a73 or later) master branch of https://github.com/tpm2-software/tpm2-pytss
+
+The terms for the systems are `remote system`, the system we want to pass the secret to and `local system`, the system which has the secret but doesn't have any TPM.
+One assumtion that will be made is that you already have the EK for the remote system on the local system, and trust it.
+
+## Concept
+The concept is that the key being activated by a call to ActivateCredential on the remote system doesn't have to be generated on the TPM, as long as the public and private parts are loaded it will succeed.
+By generating a temporary key pair on the local system we can run MakeCredential with the remote system EK, name of the locally generated key and the secret.
```

Comment:
(Using a single well-known key is better in part because it leaks nothing about the RNG on the sender side.  Though nowadays RNGs are secure enough that this doesn't matter.)

Reply:
there might be some value in using a temporary keys, but I need to think about that a bit more, but for example having a policy for the key that's tied to the TPM clock/reset counter might be useful to stop replay attacks and then one key per attestation is needed

Reply:
Since the activation key itself (private or public parts) is not used for anything at all in `Make/ActivateCredential`, and since the only part of it that matters is the policy digest set on it (it gets hashed into its name), the key can be anything at all, including the same well-known key.

If you use a temp key then you have to send its private part in cleartext, and since it's not needed all you're doing is bloating the message to the remote system.

Reply:
> but for example having a policy for the key that's tied to the TPM clock/reset counter might be useful to stop replay attacks and then one key per attestation is needed

You don't have to change the temp/well-known key for that!!  The only that's used about that key is its _name_ and the policy digest that's part of the public area that's digested to make the name.

So one WK can have many names, one per-policy.

So the message really needs to look like this: `{ "target": "<SHA256(EKpub)>", "policy_options": { ... }, "credential": ["<credentialBlob>","<secret>"], "large_ciphertext": "<...>"}`.  The policy options would be used to fill a policy template from which to compute a policy digest to use in `TPM2_LoadExternal()` when loading the WK so it matches the name used by the sender.  The `credential` would be the two outputs of `TPM2_MakeCredential()`.  And `large_ciphertext` would be a ciphertext made with the small secret as the symmetric key (and the ciphertext should include a MAC/HMAC of the ciphertext).

Reply:
That said, I wouldn't use a policy tied to the `resetCount` or time or whatever.  The point for me is to deliver things like filesystem keys needed to mount `/` and continue booting, and since those don't change, and since we don't want to have their cleartext on the server-side... there's no need to call `MakeCredential` repeatedly.  To prevent replays, instead, we'd super-encrypt with another `MakeCredential` that is tied to an AK that has `stClear`, and the inner `MakeCredential` payload would have a policy that requires a non-resettable PCR to have the initial value -- that way the recipient can immediately extend the PCR to make it impossible to re-activate the previously sent credential without an intervening reboot.

Putting it all together:

```
   Client->Server: EKpub, AKpub, eventlog, timestamp, TPM_Quote(AK, pcr_selection, timestamp)
   Server->Client: TPM2_MakeCredential(EKpub, AKname, outer_sym_key),
                   Encrypt_outer_sym_key(TPM2_MakeCredential(EKpub, WKname, inner_sym_key) ||
                                                          short_term_secrets),
                   Encrypt_inner_sym_key(long_term_secrets)
```

The policy for the WK would require a non-resettable PCR to have the initial value.

The `AKname` constructed by the server would use the `AKpub` and the attributes the server wants the `AK` to have (including `stClear`).

Reply:
I'll add a note that the key might be temporary but keep the code, I don't like long living keys outside ssh

Reply:
Here's a long lived AES key (in hex): `00000000000000000000000000000000` :)

I can make many more, though not all of them, heh.

Reply:
I still think permanent, well-known >> temporary.  Sending the private key... you might as well hard-code it.